versioned/tiered/src/main/java/com/dremio/nessie/versioned/impl/EntityType.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (C) 2020 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.dremio.nessie.versioned.impl;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import com.dremio.nessie.tiered.builder.BaseConsumer;
+import com.dremio.nessie.tiered.builder.CommitMetadataConsumer;
+import com.dremio.nessie.tiered.builder.FragmentConsumer;
+import com.dremio.nessie.tiered.builder.L1Consumer;
+import com.dremio.nessie.tiered.builder.L2Consumer;
+import com.dremio.nessie.tiered.builder.L3Consumer;
+import com.dremio.nessie.tiered.builder.RefConsumer;
+import com.dremio.nessie.tiered.builder.ValueConsumer;
+import com.dremio.nessie.versioned.impl.PersistentBase.EntityBuilder;
+import com.dremio.nessie.versioned.store.Id;
+import com.dremio.nessie.versioned.store.NotFoundException;
+import com.dremio.nessie.versioned.store.SaveOp;
+import com.dremio.nessie.versioned.store.Store;
+import com.dremio.nessie.versioned.store.StoreOperationException;
+import com.dremio.nessie.versioned.store.ValueType;
+
+final class EntityType<C extends BaseConsumer<C>, E extends PersistentBase<C>> {
+
+  static final Map<ValueType<?>, EntityType<?, ?>> byValueType = new HashMap<>();
+
+  static final EntityType<RefConsumer, InternalRef> REF = new EntityType<>(ValueType.REF, InternalRef.Builder::new);
+  static final EntityType<L1Consumer, L1> L1 = new EntityType<>(ValueType.L1, com.dremio.nessie.versioned.impl.L1.Builder::new);
+  static final EntityType<L2Consumer, L2> L2 = new EntityType<>(ValueType.L2, com.dremio.nessie.versioned.impl.L2.Builder::new);
+  static final EntityType<L3Consumer, L3> L3 = new EntityType<>(ValueType.L3, com.dremio.nessie.versioned.impl.L3.Builder::new);
+  static final EntityType<ValueConsumer, InternalValue> VALUE = new EntityType<>(ValueType.VALUE, InternalValue.Builder::new);
+  static final EntityType<FragmentConsumer, Fragment> KEY_FRAGMENT = new EntityType<>(ValueType.KEY_FRAGMENT, Fragment.Builder::new);
+  static final EntityType<CommitMetadataConsumer, InternalCommitMetadata> COMMIT_METADATA = new EntityType<>(ValueType.COMMIT_METADATA,
+      InternalCommitMetadata.Builder::new);
+
+  final ValueType<C> valueType;
+  final Supplier<C> producerSupplier;
+
+  private EntityType(ValueType<C> valueType, Supplier<C> producerSupplier) {
+    this.valueType = valueType;
+    this.producerSupplier = producerSupplier;
```

Comment:
Is a producer not a supplier? Would it be better to call this "supplier" instead and use consistently throughout?